File: cpp/878.cpp
#include<bits/stdc++.h>
class Solution {
public:
    int nthMagicalNumber(int N, int A, int B) {
        int a_b = A/gcd(A,B)*B;
        int MOD = (int)1e9 + 7;
        long i = 1, j = (long)1e15;
        long mid, n;
        while (i<j) {
            mid = (j - i)/2 + i;
            n = mid/A + mid/B - mid/a_b;
            if (n < N) {
                i = mid + 1;
            } else {
                j = mid;
            }
        }
        return (int) (i % MOD);
    }
    int gcd(int x, int y) {
        return !x ? y: gcd(y % x, x); 
    }

};

File: python/878.py
class Solution:
    def nthMagicalNumber(self, N: int, A: int, B: int) -> int:
        def gcd(x, y):
            if not x:
                return y
            else:
                return gcd(y % x, x);

        A_B = A*B//gcd(A, B)
        MOD = 10**9 + 7

        i = 1
        j = 10 ** 15
        while i < j:
            mid = (i + j) // 2
            n = mid // A + mid // B - mid // A_B
            if n < N:
                i = mid + 1
            else:
                j = mid

        return i % MOD


